### PR TITLE
contrast and gamma post-processing effect

### PR DESCRIPTION
**Pull request by mrcheko**
Don't think birghtness is needed

### CODE REVIEW

**Comment by scrawl** [comment]
I guess having it work on Windows at least would be better than nothing. IMO it can be merged like this, but perhaps the ingame options should be commented out until it's working on all platforms, to prevent mass confusion.

**Comment by kcat** [comment]
I still don't like it automatically messing with the gamma without user input, since this effects the whole system. Gamma control is an entirely system-dependent thing, as different people will have and want different values (which they'll likely already have set up in their system settings). For me, for instance, this default value of 1.34 makes my screen look washed-out (my system default is at 1, and I've used my monitor's built-in controls for fine tuning it).

**Comment by zinnschlag** [comment]
@scrawl I agree.
@kcat I fully agree.

Also, there have been multiple additions to the settings Windows recently. Can we go lighter on this until 1.1 please. We are missing proper localisation for these UI elements and we can't add them properly before 1.1.

**Comment by mrcheko** [comment]
all remarks are taken into account

**Comment by kcat** [comment]
I don't think setting the defaults to 1 is good enough since people can have other values set on their system, so you'd still end up changing it on them.

**Comment by mrcheko** [comment]
We could initialize gamma and contrast settings to values from the system but it seems that there is no way to get them. 
On the other hand we can get system gamma ramp and determine gamma from that, but in general the ramp can be affected not only by gamma but also contrast and brightness, so assuming such calculation will only account for gamma (as we can't assume anything about contrast, brightness) then the result will be fairly wrong.

**Comment by kcat** [comment]
> We could initialize gamma and contrast settings to values from the system but it seems that there is no way to get them.

Perhaps we could implement an interface with OS-specific implementations instead of using SDL for it. As nice as SDL can be for providing a platform-agnostic layer, if it doesn't allow us to get the settings we need ('gamma' and 'contrast' in this case), and we can't figure it out from the returned color tables, it becomes more of a hindrance than helpful. Perhaps an SDL feature request, to get/set the gamma and contrast settings from the system instead of having to play with the color tables directly.

Or perhaps another option. Load the color tables at start up, and keep them stored as the base. Gamma=1 and contrast=1 corresponds to this table, and increasing the gamma then increases it relative to that starting table, and same with contrast. In essence, the gamma and contrast settings become adjustments from the system settings. Though that still feels kinda kludgy, it would allow us to maintain the system setting as the default.

**Comment by psi29a** [comment]
I vote for do both.  One to fix out problem now, with a link back to the feature request.

**Comment by mrcheko** [comment]
> Perhaps we could implement an interface with OS-specific implementations instead of using SDL

Yes, that's what I was looking for when was searching for a solution. But I didn't find any even for Windows. As I understood gamma ramp is the only shared resource on the system level and gamma, contrast etc. are just app-specific parameters, so that each app can implement it's own gamma formula.
Also, returning to SDL, it has int SDL_SetGamma(float redgamma, float greengamma, float bluegamma) function, so what had prevented them to implement Get-function?

> Load the color tables at start up, and keep them stored as the base

I have already considered that solution, too. But calculations will be wrong: if say some gamma g1 was applied in NVidia/ATI driver controls so applying 2nd gamma (g2) in the game controls will finally give us: 
x^(1/g1) * x^(1/g2) = x^(1/g1+1/g2) while we wanted x^[1/(g1+g2)], where x - is color intesity from color table. So that change to color table will not be gamma.

As a side note, I checked several AAA-titles including Morrowind, and they all didn't react on system's gamma (like I did here) - even those which had only brightness control. The only exception was Assasin's Creed which actualy takes system gamma into account though it has only brightness control. So I believe they really use system's gamma ramp as base and just make base_intensity+game_brightness. But gamma is not so simple.

**Comment by mrcheko** [comment]
Also forgot to mention, when running in windowed mode they prohibit any gamma/brightness adjusting.

**Comment by kcat** [comment]
> Also, returning to SDL, it has int SDL_SetGamma(float redgamma, float greengamma, float bluegamma) function, so what had prevented them to implement Get-function?

Likely the same reason we can't. With just the color tables, there's no reliable way to figure out the gamma from whatever values they can have. Though I'm surprised the OS doesn't have a way for apps to query and set gamma and contrast values since that's usually what they expose to the user for tweaking.

> I have already considered that solution, too. But calculations will be wrong: if say some gamma g1 was applied in NVidia/ATI driver controls so applying 2nd gamma (g2) in the game controls will finally give us:
x^(1/g1) * x^(1/g2) = x^(1/g1+1/g2) while we wanted x^[1/(g1+g2)], where x - is color intesity from color table. So that change to color table will not be gamma.

You could deconstruct the value back to its gamma component, add to that, then reconstruct it. Although that presumes there's no other factors affecting the color table (like contrast or brightness). You could make assumptions, but it is a kludge like I said.

**Comment by mrcheko** [comment]
> Likely the same reason we can't. With just the color tables, there's no reliable way to figure out...

Yes, I understand this. I noticed that SDL_SetGamma because IMO if there was a possibility to get float values for gamma and contrast from the system they would implement that already.

> You could deconstruct the value back to its gamma component, add to that, then reconstruct it... but it is a kludge...

Yes, and that's why

> Load the color tables at start up, and keep them stored as the base... the gamma then increases it relative to that starting table...

is not possible. We can't deconstruct any parameter from the base. And how are you going to apply new gamma to the intesity from the base color table? Multiply base intensity by new calculated value for new gamma (if you will add - it will be even more incorrect). But as I showed above multiplication is also wrong. I just gave false explanation of 'x' - it is color intensity from standard color table (gamma = br = contrast = 1.0): 0, 256, 512, 768, ... (256 elements for each color)

To sum up we don't have a good solution to take into account system's gamma and contrast configuration. And that's why I also wrote about investigating AAA-titles behaviour.

**Comment by kcat** [comment]
> Yes, and that's why ... is not possible. We can't deconstruct any parameter from the base.

It is possible, it'll just be a bit weird if there's contrast or brightness adjustments included in the base color table.
```
float k = i/256.0;
float gamma = logf(k) / logf(BaseTable[i]/65536.0f);

k = (k-0.5f)*ContrastSetting + 0.5f;
gamma += GammaSetting;
k = powf(k, 1.0f/gamma);
new_table[i] = (ushort)clamp(k*65536.0f, 0.0f, 65535.0f);
```
If the BaseTable is derived using gamma only (i.e. no brightness and/or contrast adjustments), this will do the right thing (our gamma will add to the system gamma). If it has contrast and/or brightness adjustments, it'll behave a bit weird but still increase the power curve represented by the color table. If our gamma and contrast settings are the default 1, this will also pass through and not mess with the user settings (ignoring floating-point precision errors, which hopefully won't encroach into the 16-bit range the table uses but probably would with our luck).

> To sum up we don't have a good solution to take into account system's gamma and contrast configuration.

We don't, but unfortunately that still leaves us with the problem of messing around with the system's gamma and contrast settings without the user's permission.
